File: renderer.py
import os
import uuid
from playwright.async_api import async_playwright

# Output directory
OUTPUT_DIR = os.path.join(os.path.dirname(__file__), "output")
os.makedirs(OUTPUT_DIR, exist_ok=True)


async def render_thumbnail(html_content: str, output_filename: str = None) -> str:
    """
    Render HTML content to a PNG screenshot using Playwright headless Chromium.

    Args:
        html_content: Full HTML string to render
        output_filename: Optional filename for the output PNG. If None, a UUID is generated.

    Returns:
        Absolute path to the generated PNG file.
    """
    if output_filename is None:
        output_filename = f"thumbnail_{uuid.uuid4().hex[:8]}.png"

    output_path = os.path.join(OUTPUT_DIR, output_filename)

    async with async_playwright() as p:
        browser = await p.chromium.launch(headless=True)
        page = await browser.new_page(
            viewport={"width": 1920, "height": 1080},
        )

        # Load HTML content directly
        await page.set_content(html_content)

        # Wait for Google Fonts and all images to load
        await page.wait_for_load_state("networkidle")

        # Extra wait to ensure font rendering is complete
        await page.wait_for_timeout(1000)

        # Screenshot only the .thumbnail element
        thumbnail_element = await page.query_selector("#thumbnail")
        if thumbnail_element:
            await thumbnail_element.screenshot(path=output_path, type="png")
        else:
            # Fallback: screenshot the full page
            await page.screenshot(
                path=output_path,
                type="png",
                clip={"x": 0, "y": 0, "width": 1920, "height": 1080},
            )

        await browser.close()

    return output_path

File: src/services/renderer.js
const { chromium } = require('playwright');
const config = require('../config');

/**
 * Browser Pool - Persistent browser instance for thumbnail rendering.
 *
 * Instead of launching a new browser per request (like the Python version),
 * we keep a single browser instance and create new BrowserContexts per request.
 * This saves ~1-2s per request.
 */
let browser = null;

/**
 * Initialize the browser instance. Call once at startup.
 */
async function init() {
  if (browser) return;

  browser = await chromium.launch({
    headless: config.browserHeadless,
  });

  console.log('[Renderer] Browser pool initialized');
}

/**
 * Render HTML content to a PNG buffer using Playwright.
 *
 * @param {string} htmlContent - Full HTML string to render
 * @returns {Promise<Buffer>} PNG image as a Buffer
 */
async function renderThumbnail(htmlContent) {
  if (!browser) {
    throw new Error('Browser not initialized. Call renderer.init() first.');
  }

  const context = await browser.newContext({
    viewport: config.viewport,
  });

  try {
    const page = await context.newPage();

    // Load HTML content directly
    await page.setContent(htmlContent);

    // Wait for Google Fonts and all images to load
    await page.waitForLoadState('networkidle');

    // Extra wait to ensure font rendering is complete
    await page.waitForTimeout(config.fontLoadWait);

    // Screenshot only the #thumbnail element
    const thumbnailElement = await page.$('#thumbnail');

    let buffer;
    if (thumbnailElement) {
      buffer = await thumbnailElement.screenshot({ type: 'png' });
    } else {
      // Fallback: screenshot with clip
      buffer = await page.screenshot({
        type: 'png',
        clip: { x: 0, y: 0, ...config.viewport },
      });
    }

    return buffer;
  } finally {
    await context.close();
  }
}

/**
 * Check if the browser is alive and responsive.
 * @returns {Promise<boolean>}
 */
async function isHealthy() {
  if (!browser) return false;
  try {
    // Try creating and closing a context to verify browser is responsive
    const ctx = await browser.newContext();
    await ctx.close();
    return true;
  } catch {
    return false;
  }
}

/**
 * Gracefully shut down the browser. Call on process exit.
 */
async function shutdown() {
  if (browser) {
    await browser.close();
    browser = null;
    console.log('[Renderer] Browser pool shut down');
  }
}

module.exports = { init, renderThumbnail, isHealthy, shutdown };
